### PR TITLE
Dfs refactor

### PR DESCRIPTION
**Pull request by ecomodeller**
* Item selection by either names or number using the same `items` argument
* filename supplied to initializer instead of in `read` method
* `write` and `create` methods merged to one method
* common base class for Dfs1, Dfs2 & Dfs3

### CODE REVIEW

File: mikeio/dfs.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+)
+
+class Dfs123:
+
+    _projstr = None
+    _start_time = None
+
+    def __init__(self, filename = None):
+        self._filename = filename
+        # TODO self._read_header(filename)
```
Comment (rhaDHI) [inline mikeio/dfs.py:17]: Call this?

File: mikeio/dfs.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+    def _read_header(self, dfs):
+        self._n_items = safe_length(dfs.ItemInfo)
+        self._items = get_item_info(dfs, list(range(self._n_items)))
+        self._start_time = from_dotnet_datetime(dfs.FileInfo.TimeAxis.StartDateTime)
+        self._n_timesteps = dfs.FileInfo.TimeAxis.NumberOfTimeSteps
+        #self._timestep_in_seconds = dfs.FileInfo.TimeAxis.TimeStepInSeconds
```
Comment (rhaDHI) [inline mikeio/dfs.py:24]: Remove?

File: mikeio/dfs0.py [hunk 3]
```diff
@@ -15,71 +16,80 @@
...
+        """
+        self._filename = filename
+
+    def read(self, items=None, time_steps=None):
         """
         Read data from a dfs0 file.
 
         Parameters
         ----------
-        filename: str
```
Comment (rhaDHI) [inline mikeio/dfs0.py:30]: Doc strings removed? Should stay.

File: mikeio/dfs0.py [hunk 4]
```diff
@@ -222,8 +224,8 @@ def create(
         if datetimes is None:
             equidistant = True
 
-            if not type(start_time) is datetime:
-                raise Warning("start_time must be of type datetime.")
+            # if not type(start_time) is datetime:
```
Comment (rhaDHI) [inline mikeio/dfs0.py:227]: Remove outcommented code

File: mikeio/dfs0.py [hunk 5]
```diff
@@ -279,6 +287,7 @@ def create(
 
         delete_value = dfs.FileInfo.DeleteValueFloat
 
+        data = data.copy()
         for i in range(n_items):
             d = data[i].copy()
             d[np.isnan(d)] = delete_value
```
Comment (rhaDHI) [inline mikeio/dfs0.py:293]: Alternative used previously
```suggestion
        nan_indices = np.isclose(data, self._dfs.FileInfo.DeleteValueFloat, atol=1e-36)
        data[nan_indices] = np.nan
```

File: mikeio/dfs1.py [hunk 6]
```diff
@@ -1,3 +1,4 @@
+import warnings
```
Comment (rhaDHI) [inline mikeio/dfs1.py:1]: Not used import

File: mikeio/dfs0.py [hunk 7]
```diff
@@ -289,7 +298,7 @@ def create(
 
         dfs.Close()
 
-    def to_dataframe(self, filename, unit_in_name=False, round_time='s'):
+    def to_dataframe(self, unit_in_name=False, round_time="s"):
         """
         Read data from the dfs0 file and return a Pandas DataFrame.
```
Comment (rhaDHI) [inline mikeio/dfs0.py:304]: Update doc strings here

File: mikeio/dfs0.py [hunk 8]
```diff
@@ -15,71 +16,80 @@
...
 from .dotnet import to_dotnet_array, to_dotnet_datetime, from_dotnet_datetime
-from .dutil import Dataset, find_item
+from .dutil import Dataset, find_item, get_valid_items_and_timesteps
 from .eum import TimeStep, EUMType, EUMUnit, ItemInfo
 from .helpers import safe_length
 
 
 class Dfs0:
-    def read(self, filename, item_numbers=None, item_names=None):
+    def __init__(self, filename=None):
```
Comment (rhaDHI) [inline mikeio/dfs0.py:25]: Shouldn't it be file_path and not filename across the code?

**Comment by rhaDHI** [comment]
Is the create time series example in the readme still valid?

**Comment by rhaDHI** [comment]
There are two spelling mistakes in the readme: 
dependancies -> dependencies
meteorlogical -> meteorological

File: mikeio/dfs0.py (hunk 8 above)
Comment (ecomodeller) [reply mikeio/dfs0.py:25]: filename is a path to a filename, I have checked some other python libraries and they use file, filename, filepath to mean the same thing, I think we will stick with filename.

File: mikeio/dfs.py (hunk 1 above)
Comment (ecomodeller) [reply mikeio/dfs.py:17]: Yes, there is still some cleaning up to do...

**Comment by ecomodeller** [comment]
> There are two spelling mistakes in the readme:
> dependancies -> dependencies
> meteorlogical -> meteorological

:-) I will update the readme

File: README.md [hunk 9]
```diff
@@ -63,22 +63,19 @@ Items:
...
 >>>  from mikeio import Dfs0
 >>>  data = []
 >>>  d = np.random.random([100])
 >>>  data.append(d)
 >>>  dfs = Dfs0()
->>>  dfs.create(filename='simple.dfs0',
->>>            data=data,
->>>            start_time=datetime(2017, 1, 1),
->>>            dt=60)
+>>>  dfs.write('simple.dfs0', data, start_time=datetime(2017, 1, 1), dt=60)
```
Comment (jsmariegaard) [inline README.md:75]: maybe make the example more concise by:


    >>>  data = [np.random.random([100])]
    >>>  dfs = Dfs0()
    >>>  dfs.write('simple.dfs0', data, start_time=datetime(2017, 1, 1), dt=60)
?

Comment (ecomodeller) [reply README.md:73]: Good idea, done

**Comment by ecomodeller** [comment]
I think this it it.

:shipit: ?

File: mikeio/dfs0.py [hunk 10]
```diff
@@ -278,25 +286,21 @@ def create(
...
+        data = np.array(data)
+        data[np.isnan(data)] = delete_value
+        data_to_write = to_dotnet_array(data.T)
         t_seconds = [(t - datetimes[0]).total_seconds() for t in datetimes]
         Dfs0Util.WriteDfs0DataDouble(dfs, t_seconds, data_to_write)
 
         dfs.Close()
 
-    def to_dataframe(self, filename, unit_in_name=False, round_time='s'):
+    def to_dataframe(self, unit_in_name=False, round_time="s"):
```
Comment (rhaDHI) [inline mikeio/dfs0.py:298]: This function is not done. Unused variable 't' and input unit_in_name

File: mikeio/dfs1.py [hunk 11]
```diff
@@ -58,20 +75,21 @@ def read(self, filename, item_numbers=None, item_names=None):
         nt = dfs.FileInfo.TimeAxis.NumberOfTimeSteps
         if nt == 0:
             raise Warning("Static dfs1 files (with no time steps) are not supported.")
```
Comment (rhaDHI) [inline mikeio/dfs1.py:77]: In general: why use warnings and not exceptions?

File: mikeio/dfs0.py [hunk 12]
```diff
@@ -165,18 +158,18 @@ def _to_dfs_datatype(dtype):
...
-            data,
-            start_time=None,
-            timeseries_unit=TimeStep.SECOND,
-            dt=1.0,
-            datetimes=None,
-            items=None,
-            title=None,
-            data_value_type=None,
-            dtype=None,
+    def write(
```
Comment (rhaDHI) [inline mikeio/dfs0.py:161]: This method is too long. Please consider shortening with appropriate functions

**Pull request merged**